[TT-Train] Add generic inference example

abustamanteTT opened the pull request:
### Problem description
Previous TT-Train overloaded the functionality of nanoGPT example to include training and inference of arbitrary models.

### What's changed
This PR adds a generic LLM inference example in both C++ and Python, which allows users to perform inference with any arbitrary (currently-supported) LLM

### Checklist
https://github.com/tenstorrent/tt-metal/actions/runs/18236304518
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] New/Existing tests provide coverage for changes

athompsonTT committed:
tt-train: move ModuleBase from autograd to modules

athompsonTT committed:
add nanobind _ttml module

athompsonTT committed:
tt-train: set PIC to TRUE

athompsonTT committed:
add example script

athompsonTT committed:
update script, add includes, experiment with script bindings

athompsonTT committed:
troubleshooting refactor

athompsonTT committed:
fix troubleshooting refactor

athompsonTT committed:
support/use specifying type in to_numpy, convert all numpy tensors to float32

dmakoviichuk-tt committed:
main change is to return shared_ptr tensor

athompsonTT committed:
fixup

athompsonTT committed:
refactor nb_util

athompsonTT committed:
fix make_numpy_tensor by omitting strides

athompsonTT committed:
use TT_METAL_HOME in linear_regression python example

athompsonTT committed:
support specifying type in from_numpy, fix AutoContext get_instance

athompsonTT committed:
add row_major arg to from_numpy

dmakoviichuk-tt committed:
mnist

athompsonTT committed:
make row_major a layout enum, swap positions with new_type

athompsonTT committed:
add llama, mlp models

athompsonTT committed:
add gpt2

athompsonTT committed:
add newline

athompsonTT committed:
remove unneeded bindings

athompsonTT committed:
add call operators for new models

dmakoviichuk-tt committed:
more beautiful code

dmakoviichuk-tt committed:
Merge branch 'athompson/ttml_nanobind' of github.com:tenstorrent/tt-metal into athompson/ttml_nanobind

athompsonTT committed:
remove unused include, add mlp config default ctor

dmakoviichuk-tt committed:
Merge branch 'athompson/ttml_nanobind' of github.com:tenstorrent/tt-metal into athompson/ttml_nanobind

dmakoviichuk-tt committed:
updated notebook

athompsonTT committed:
fix build

athompsonTT committed:
remove create functions, add rw fields

athompsonTT committed:
same treatment for py_mlp_params, fix create bindings

athompsonTT committed:
add args to linear_layer ctor bindings

dmakoviichuk-tt committed:
small fixes

dmakoviichuk-tt committed:
updates

dmakoviichuk-tt committed:
not very beautiful but works

dmakoviichuk-tt committed:
right params

dmakoviichuk-tt committed:
Merge branch 'main' into athompson/ttml_nanobind

dmakoviichuk-tt committed:
small fixes

dmakoviichuk-tt committed:
fixed bf16

athompsonTT committed:
Add GPT2TransformerConfigExperimental nanobind class to module

athompsonTT committed:
nb_util: codify exception strings, cleanup

athompsonTT committed:
Merge remote-tracking branch 'origin/main' into athompson/ttml_nanobind

athompsonTT committed:
restore bfloat16 support, use correct tensor for spec/strides access

athompsonTT committed:
clarify function, type names, add comments, disallow bf16 numpy types

athompsonTT committed:
auto const->const auto, add comment about stride data

athompsonTT committed:
add/use TT_METAL_HOME env var in jupyter files

athompsonTT committed:
fix ttml path in linear regression example

athompsonTT committed:
move models out of modules, update examples

athompsonTT committed:
nb_export_enum: remove export_values call

athompsonTT committed:
Add test_autograd.py

athompsonTT committed:
fix comment typo

athompsonTT committed:
fix includes, add/use throw_exception, NB_THROW, NB_COND_THROW

athompsonTT committed:
test_autograd.py: add/organize test scenarios, check for expected exceptions

athompsonTT committed:
remove unused include, wrap assert in try/except

athompsonTT committed:
Merge remote-tracking branch 'origin/main' into athompson/ttml_nanobind

athompsonTT committed:
Merge remote-tracking branch 'origin/main' into athompson/ttml_nanobind

athompsonTT committed:
remove duplicate test cases

athompsonTT committed:
remove debug code, update asserts for clarity

athompsonTT committed:
re-append ttml dir to import path

athompsonTT committed:
update nanobind cmake, add STABLE_API, NB_SHARED, LTO options

athompsonTT committed:
nb_autograd: add named args

athompsonTT committed:
move things into submodules, add named args, fix notebooks

*also comment out 'sum' unary op as it is not implemented

athompsonTT committed:
linear_regression.py: add copyright heading

athompsonTT committed:
nb_optimizers.cpp: #include nanobind/stl/variant.h

*Needed for get_state_dict return type

athompsonTT committed:
move export_enum, make_[metal,numpy]_tensor to ttml::nanobind namespace

athompsonTT committed:
move bindings into ttml::nanobind namespace

athompsonTT committed:
nb_models.cpp: scope module objects

athompsonTT committed:
linear_regression.py: fix submodule path

athompsonTT committed:
remove TT_METAL_HOME cell in notebooks

athompsonTT committed:
fix headings

athompsonTT committed:
register __call__ binding on base class

athompsonTT committed:
register binary ops in autograd::Tensor bindings

athompsonTT committed:
register ops directly instead of using lambdas

athompsonTT committed:
Merge remote-tracking branch 'origin/main' into athompson/ttml_nanobind

athompsonTT committed:
remove __init__.hpp as it is effectively empty

athompsonTT committed:
correct typeo, remove more duplicate tests

athompsonTT committed:
update examples and test to use tt-train build path

athompsonTT committed:
add install targets for _ttml

athompsonTT committed:
make ttml 'module', have it find and import _ttml

abustamanteTT committed:
Recommitting on different branch

athompsonTT committed:
update ttml module to explicitly search tt-train first

athompsonTT committed:
return false at bottom of try_setup_ttml_so function in ttml module

abustamanteTT committed:
Added sampling

athompsonTT committed:
ttml module: move imports into function body

athompsonTT committed:
Merge remote-tracking branch 'origin/main' into athompson/ttml_nanobind

athompsonTT committed:
remove get/set_generator bindings

athompsonTT committed:
Merge branch 'main' into athompson/ttml_nanobind

abustamanteTT committed:
Fixed CMakeLists, main.cpp
Continued inference notebook

abustamanteTT committed:
Merge branch 'athompson/ttml_nanobind' into abustamante/inference_example

abustamanteTT committed:
Updated sampling op nanobind default argument
Updated sampling op signature to have nullptr default

abustamanteTT committed:
Working on set_gradient_mode

abustamanteTT commented on tt-train/sources/examples/llm_inference/llm_inference.ipynb:
```diff
@@ -0,0 +1,245 @@
+    "else:\n",
+    "    logits_mask_tensor = None\n",
+    "\n",
+    "prompt_str = \"The difference between cats and dogs is:\"\n",
```
This prompt is hard-coded to allow rapid iteration/debugging; will replace with user input

abustamanteTT committed:
Added nanobind filesystem header to nb_models.cpp

abustamanteTT committed:
Added safetensors loading

abustamanteTT committed:
Removed debug print from main.cpp

abustamanteTT committed:
Added numpy randint

abustamanteTT committed:
Fixed weight tying, set temp to 0

abustamanteTT committed:
Set temperature to 0, enabled weight tying

abustamanteTT committed:
Merge remote-tracking branch 'origin/main' into abustamante/inference_example

abustamanteTT committed:
Merge remote-tracking branch 'origin/main' into abustamante/inference_example

dmakoviichuk-tt commented on tt-train/sources/examples/llm_inference/llm_inference.ipynb:
```diff
@@ -0,0 +1,736 @@
+   "outputs": [],
+   "source": [
+    "tokenizer = AutoTokenizer.from_pretrained(\"gpt2\")\n",
+    "transformer_cfg = safe_load(open(\"tt-train/configs/training_shakespeare_gpt2s.yaml\", \"r\"))[\"training_config\"][\"transformer_config\"]"
```
lets remove "tt-train" from path

dmakoviichuk-tt commented on tt-train/sources/examples/llm_inference/llm_inference.ipynb:
```diff
@@ -0,0 +1,736 @@
+      "    Dropout probability: 0.2\n",
+      "    Num blocks: 12\n",
+      "    Positional embedding type: Trainable\n",
+      "    Runner type: Default\n",
```
should be ok to use default runner type it impacts backward pass only

dmakoviichuk-tt reviewed (approved):
Please cleanup your jupyter notebook before merge a nd fix a few comments.

abustamanteTT committed:
Cleaned up llm_inference.ipynb

abustamanteTT committed:
Fixed some of the separators

abustamanteTT committed:
Fixed pytorch loading to use BF16

dmakoviichuk-tt commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,169 @@
+
+    std::string model_path = "configs/training_shakespeare_gpt2s";
+    std::string tokenizer_path =
+        "/home/ubuntu/.cache/huggingface/hub/models--gpt2/snapshots/607a30d783dfa663caf39e06633721c8d4cfcd7e/"
```
dont forget to remove hardcoded path.

abustamanteTT committed:
Added model->eval()
Added seed CLI option
Removed hardcoded paths

abustamanteTT committed:
Output seed in C++

abustamanteTT committed:
Added seed as configurable global in llm_inference.ipynb

abustamanteTT committed:
Seed

abustamanteTT committed:
Removed PyTorch inference

athompsonTT commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,169 @@
+            start_idx = static_cast<uint32_t>(prompt_tokens.size() - max_sequence_length);
+        }
+        // Fill padded array
+        for (uint32_t i = 0; i < max_sequence_length; ++i) {
```
`std::fill(prompt_tokens_padded.begin(), std::advance(prompt_tokens_padded.begin(), max_sequence_length), 0);` perhaps

abustamanteTT replied:
I think even a `std::fill(prompt_tokens_padded.begin(), prompt_tokens_padded.end(), 0);`?

abustamanteTT committed:
Merge remote-tracking branch 'origin/main' into abustamante/inference_example

abustamanteTT committed:
Swapped for-loops for STL functions per PR comments

abustamanteTT committed:
Dead code prune

abustamanteTT committed:
Moved config.py to ttml/common
Updated llm_inference.ipynb to use config.py
Updated transformers example to use ttml.common.config

abustamanteTT committed:
Addressed Copilot PR coments

dmakoviichuk-tt commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,166 @@
+        ttml::core::from_vector(mask, ttnn::Shape({1, 1, max_sequence_length, max_sequence_length}), device));
+
+    // Create a large negative mask for out-of-vocab logits
+    auto vocab_mask = std::vector<float>(padded_vocab_size - original_vocab_size, 1e4F);
```
please add check if you need to apply padding_mask

dmakoviichuk-tt replied:
Same in python

abustamanteTT replied:
Added check to C++, already existed in Python, commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

rfurko-tt commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,166 @@
+using MeshShape = tt::tt_metal::distributed::MeshShape;
+using Model = std::shared_ptr<ttml::models::BaseTransformer>;
+
+ttml::autograd::TensorPtr run_model(
```
lets remove run_model method, we don't need it anymore

abustamanteTT replied:
Commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

rfurko-tt commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,166 @@
+    std::string prompt = "";
+    std::string safetensors_path = "";
+    float temperature = 0.0F;
+    uint32_t tokens_to_generate = 256;
```
256U

abustamanteTT replied:
Commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

rfurko-tt commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,166 @@
+    app.add_option("--safetensors", safetensors_path, "Path to the model safetensors file")
+        ->default_val(safetensors_path);
+    app.add_option("--tokenizer", tokenizer_path, "Path to the tokenizer files")->default_val(tokenizer_path);
+    app.add_option("--temperature", temperature, "Sampling temperature")->default_val(temperature);
```
isn't it part of the yaml config?

abustamanteTT replied:
It is, I hadn't wanted to copy code from the nanoGPT example, but I'm noticing that we don't have a unified EvalConfig struct + parsing function. I'll copy those over from nanoGPT, and look into creating a generic config class

abustamanteTT replied:
Updated with code from nanoGPT in commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa), long-term I think a generic config class would be good

rfurko-tt commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,166 @@
+
+    // Load model
+    auto yaml_config = YAML::LoadFile(model_path + ".yaml");
+    auto config = ttml::models::gpt2::read_config(yaml_config["training_config"]["transformer_config"]);
```
does it only support gpt2 model?

abustamanteTT replied:
Long term, no, short term I think we only have GPT2 and LLama safetensors

rfurko-tt commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,166 @@
+    app.add_option("-p,--prompt", prompt, "Prompt for text generation")->default_val(prompt);
+    app.add_option("--safetensors", safetensors_path, "Path to the model safetensors file")
+        ->default_val(safetensors_path);
+    app.add_option("--tokenizer", tokenizer_path, "Path to the tokenizer files")->default_val(tokenizer_path);
```
if it's not provided here -> will it use tokenizer from config?

abustamanteTT replied:
It will now! Commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

rfurko-tt commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,166 @@
+        prompt_tokens_padded, ttnn::Shape({1U, 1U, 1U, max_sequence_length}), device, ttnn::Layout::ROW_MAJOR));
+
+    // Generate tokens
+    std::cout << "Generated text:\n";
```
my understanding that it answers only a single prompt and closes program. @abustamanteTT  @dmakoviichuk-tt is it expected? 

in `nano_gpt/main.cpp` i was doing while true logic :D

rfurko-tt commented on tt-train/sources/examples/python/transformers/training.py:
```diff
@@ -13,11 +13,11 @@
 sys.path.append(f'{os.environ["TT_METAL_HOME"]}/tt-train/sources/ttml')
 
 import ttml
+from ttml.common.config import get_config, DeviceConfig, TrainingConfig
```
do we only want to reuse config, nothing else?

abustamanteTT replied:
Just did a refactor, I think there's more in the other files to reuse, but I'd need to take another look. Commit [544d3da](https://github.com/tenstorrent/tt-metal/pull/29704/commits/544d3dae1c39fb0442c00d2386348248082532aa)

rfurko-tt commented on tt-train/tests/python/test_autograd.py:
```diff
@@ -6,6 +6,7 @@
 import sys
 
+
 # sys.path.append(f'{os.environ["TT_METAL_HOME"]}/tt-train/build/sources/ttml')
```
i know it's not part of your PR, but could you please remove comment :)

abustamanteTT committed:
Addressed PR comments
Refactored llm_inference.ipynb
Reorganized ttml.common

abustamanteTT committed:
Removed comment by request :)

abustamanteTT committed:
Updated causal mask generation in main.cpp

abustamanteTT commented on tt-train/sources/examples/llm_inference/main.cpp:
```diff
@@ -0,0 +1,166 @@
+
+    // Build mask (causal) for attention
+    std::vector<float> mask;
+    mask.reserve(static_cast<size_t>(max_sequence_length * max_sequence_length));
```
I like this suggestion, I think it makes it much more readable too. Commit [d2369dd](https://github.com/tenstorrent/tt-metal/pull/29704/commits/d2369dd3893f0c36d7c433108d9dd0e41ac29ed2)

dmakoviichuk-tt reviewed (approved):
we can cleanup inference example a little bit more.

abustamanteTT committed:
Addressed final PR comments
Reverted causal mask generation because of issues

abustamanteTT committed:
Merge branch 'main' into abustamante/inference_example

abustamanteTT committed:
Removed unused variable warning

abustamanteTT committed:
Merge branch 'abustamante/inference_example' of github.com:tenstorrent/tt-metal into abustamante/inference_example
